Expands on the declarative alternative solution for Web Install (#975)
MicrosoftEdge/MSEdgeExplainers: WebInstall/explainer.md

### Declarative install

> An alternate solution is to have a declarative way to install web apps. This can be achieved by allowing a new `target` type of `_install` to the HTML anchor tag. It can also use the [`rel`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel) attribute to hint to the UA that the url in the link should be installed

Reviewer: ```suggestion
An alternate solution is to have a declarative way to install web apps. This can be achieved by allowing a new `target` type of `_install` to the HTML anchor tag. It can also use the [`rel`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel) attribute to hint to the UA that the url in the link should be installed.
```

* Platform fallback to navigate to the content automatically.
* Does not need JavaScript.

*Cons:*
* Takes the user out of the current context, providing no alternative if the use case benefits from them staying in context.
* Limits the amount of information a developer can act upon that the promise provides, such as if the installation was successful.
* Developers can't easily detect UA declarative support in order to be able to tailor their UX to different situations.
* No support for the concept of an `install_url` (requires additional work to the HTML specification).  Using an `install_url` as the `href` could confusingly land the user on a seemingly blank page for UA's that don't support declarative install.
* No clear way to pass additional information intended for the UA (like the referral-info).
* Longer `a` tags: if we address the missing `install_url` and `manifest_id` parameters and decided to keep the `href` as a link then we could be potentially introducing lengthier and less readable tags similar to: `<a href="https://airhorner.com" rel="install" target="_blank" installUrl="https://airhorner.com/install.html" manifestId="https://airhorner.com/airhornApp">honk</a>`.

> * More complex combinations for the UA to take into account: additional attributes that act on a link HTML tag (`a`) like the target mean there is an increased set of scenarios that might have unintended consequences for end users. For example, how does a target of `_ top` differ from `_blank`? While we could look at ignoring the `target` attribute if a `rel` attribute is present, the idea is to use acquisition mecanisms that are already present in UAs. 

Reviewer: ```suggestion
* More complex combinations for the UA to take into account: additional attributes that act on a link HTML tag (`a`) like the target mean there is an increased set of scenarios that might have unintended consequences for end users. For example, how does a target of `_ top` differ from `_blank`? While we could look at ignoring the `target` attribute if a `rel` attribute is present, the idea is to use acquisition mechanisms that are already present in UAs. 
```

Having stated this, we believe that a declarative implementation is a simple and effective solution, and a future entry point for the API. It should be [considered for a v2](#future-work) of the capability. For the current solution, **we've decided to go with an imperative implementation since it allows more control over the overall installation UX**:
* Allows the source to detect if an installation occurred with ease. (resolves/rejects a promise).
* Supports `install_url`. This url can be an optimized url or the normal homepage that an application already has. The advantage is that unlike a declarative version, there is no scenario where an end user can be left stranded accidentally in a blank page that is meant to be a lightweight entry point to the app.

> * Code can be used to detect if an origin has [permission](#new-installation-permission-for-origin) to install apps, and UX can be tailored to change accordingly (for example, remove a button or display a link instead).

Reviewer: ```suggestion
* Code can be used to detect if an origin has [permission](#new-installation-permission-for-origin) to install apps, as well as if the UA supports the API, and UX can be tailored to change accordingly (for example, remove a button or display a link instead).
* ```

* The developer ergonomics of handling a promise are better than responding to an `a` tag navigation.

> * Keeps the user in the context, which *can* be beneficial in certain scenarios (if the developer *wants* to take the user out of the current context, they *can* do so with a `a` tag).

Reviewer: In the TAG review, some of the feedback was this: "'Takes the user out of the current context.' is a downside, but the TAG currently disagrees that that's a downside and would need to be convinced. What data have you collected that shows that it's bad for users to have to visit a website before installing it? A "peek" UI like [https://resources.arc.net/hc/en-](https://resources.arc.net/hc/en-us/articles/19335302900887-Peek-Preview-Sites-From-Pinned-Tabs)"

You and I are in agreement that there exist potential dev benefits to allow the current context to be retained, but I don't think you've managed to provide an actual example here to demonstrate that.  I just think we need to flush this point out a little further.  They ask what data we have for our suggestion.  We have at least one web developer (store.app) who's desired model would be to have the install occur without navigating away (in theory to facilitate multiple installations).  TAG doesn't present data to the contrary.  I don't think we necessarily need data, but flushing out realistic scenarios might help close this part of the conversation down.

### Declarative install

> **An alternate solution to allow installation of web apps is by allowing a new target type of `_install` to the HTML anchor tag.**

Author: both can be alternatives tbf. Depends on how it is designed.

`<a href="https://airhorner.com" target="_install">honk</a>`

Pros:
* Platform fallback to navigate to the content automatically.
* Does not need JavaScript.

Cons:
* Takes the user out of the current context, providing no alternative if the use case benefits from them staying in context.

> * Limits the amount of information a developer can act upon that the promise provides, such as if the prompt was shown.

Reviewer: ```suggestion
* Limits the amount of information a developer can act upon that the promise provides, such as if the installation was successful.
```

> * Developers can't modify their UX to tailor to situations where the origin permission isn't granted.

Reviewer: ```suggestion
* Developers can't easily detect UA declarative support in order to be able to tailor their UX to different situations.
```

> * No support for the concept of an `install_url` (requires additional work to the HTML specification).

Reviewer: ```suggestion
* No support for the concept of an `install_url` (requires additional work to the HTML specification).  Using an `install_url` as the `href` could confusingly land the user on a seemingly blank page for UA's that don't support declarative install.
```

> * Can't easily pass additional information like the referral-info.

Reviewer: ```suggestion
* No clear way to pass additional information intended for the UA (like the referral-info).
```

**Another alternate solution based in a declarative approach could see the use of the [rel](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel) attribute to hint to the UA that the url in the link should be installed.**

`<a href="https://airhorner.com" rel="install">honk</a>`

Pros (same as above):
* Basic platform default fallback (navigation).
* Does not need JavaScript.

Cons (same as above plus): 

> * Unintended navigation to `install_url`: If a UA doesn't implement the web install feature, and the `install_url` is used in the `href` value, it may fallback to navigate to that `install_url`, potentially leaving the end user in a blank document. (this could be resolved by creating more attributes to the `a` tag to define the additional parameters that the feature requires).

Reviewer: Unintended navigation would still be an issue with the `target` solution above.  same with the problem with potentially needing additional new attributes to handle the other inputs (`install_url` and `manifestId`).

In general, I think that declarative install section can be simplified.  I suggest that you present both options (`target` and `rel`) first and call out all of the common issues once (as opposed to twice). Then you can additionally call out the additional complexity of how to handle an `install` `rel` with varying `target` values.

The overall message content you have in this update is sound.  I just suggest minimizing the duplication to make this easier to read/follow.

* More complex combinations for the UA to take into account: additional attributes that act on a link HTML tag (`a`) like the target mean there is an increased set of scenarios that might have unintended consequences for end users. For example, how does a target of `_ top` differ from `_blank`? While we could look at ignoring the `target` attribute if a `rel` attribute is present, the idea is to use acquisition mecanisms that are already present in UAs. 

Having stated this, we believe that a declarative implementation is a simple and effective solution, and a future entry point for the API. It should be [considered for a v2](#future-work) of the capability. For the current solution, **we've decided to go with an imperative implementation since it allows more control over the overall installation UX**:
* Allows the source to detect if an installation occurred with ease. (resolves/rejects a promise).

> * Supports `install_url`. This url can be an optimized url or the normal homepage that an application already has. The advantage is that unlike a declarative version, there is no scenario where an end user can be left stranded accidentally in a blank page that is meant to be a  lightweight entry point to the app.

Reviewer: ```suggestion
* Supports `install_url`. This url can be an optimized url or the normal homepage that an application already has. The advantage is that unlike a declarative version, there is no scenario where an end user can be left stranded accidentally in a blank page that is meant to be a lightweight entry point to the app.
```
